workflow/libero_flow_poc/libero_flow/scheduler.py

```diff
@@ -0,0 +1,231 @@
+    :param input_data: Dict
+    :return:
+    """
+    # get workflow definition
+    loader = FlowLoader()
+    workflow_def = loader.get_workflow(name)
+
+    if workflow_def:
+        workflow_payload = {
+            "name": workflow_def.get('name'),
+            "input_data": json.dumps(input_data),
+            "config": json.dumps(workflow_def.get('config'))
+        }
+
+        # create workflow
+        response = requests.post(WORKFLOW_API_URL, data=workflow_payload)
+        workflow_id = response.json().get('instance_id')
+
+        if workflow_id:
+            # create activities
+            for activity in workflow_def.get('activities', []):
+                activity_payload = copy.deepcopy(activity)
+                activity_payload['config'] = json.dumps(activity_payload['config'])
+                activity_payload['workflow'] = workflow_id
+
+                requests.post(ACTIVITY_API_URL, data=activity_payload)
+
+        update_workflow_status(workflow_id=workflow_id, status='In Progress')
+
+        schedule_decision(workflow_id=workflow_id)
+
+        # TODO if successful then send workflow created event
+
+    else:
+        # TODO if unsuccessful then send workflow creation failed event
+        print(f'No Workflow definition found for {name}')
+
+
+def activity_result_handler(data: Dict[str, Any]) -> None:
+    print(f'[x] Activity results received: {data}')
+
+    result = data['data']
+
+    activity_state = get_activity_state(result['activity_id'])
+    update_activity_status(activity_id=result['activity_id'], status=result['result'])
+    schedule_decision(workflow_id=activity_state['workflow'])
+
+
+def decision_result_handler(data: Dict[str, Any]) -> None:
+    print(f'[x] Decision results received: {data}')
+
+    decision = data['data']
+
+    if decision['decision'] == 'start-workflow':
+        create_workflow(data['name'], data['input_data'])
+
+    elif decision['decision'] == 'schedule-activities':
+        for activity in decision['activities']:
+            # TODO need to check activity does not currently have an active timeout
+            schedule_activity(activity['instance_id'])
+
+    elif decision['decision'] == 'workflow-finished':
+        update_workflow_status(workflow_id=decision['workflow_id'], status=FINISHED)
+
+    elif decision['decision'] == 'workflow-failure':
+        update_workflow_status(workflow_id=decision['workflow_id'], status=FINISHED)
+
+    elif decision['decision'] == 'do-nothing':
+        pass
+
+
+def workflow_starter(data: Dict[str, Any]) -> None:
+    print(f'[x] Workflow starter received: {data}')
+    create_workflow(data['name'], data['input_data'])
+
+
+@setup_exchanges_and_queues
+def main():
+    with get_channel() as channel:
+        print('Scheduler running...')
+        print(' [*] Waiting for Messages. To exit press CTRL+C')
+
+        channel.basic_consume(message_handler(activity_result_handler), queue=ACTIVITY_RESULT_QUEUE, no_ack=False)
```

Comment:
are these 3 calls concurrent? Or running one after the other?

Reply:
Would seem strange to e.g. waiting on activity results while there are other decision results to show. This is probably just a debugging method however.

Reply:
They are executed on one thread (for this POC) but are just callback assignments. Meaning you can receive messages from all 3 queues independently, but I can only execute one callback at a time.

As the callback functions don't hold state it should not matter, you could implement multi threading/processing or start x2 schedulers and you could then execute two callbacks at a time etc.